### PR TITLE
fix: catch error to find a locator

### LINKED ISSUE #3777: [Bug?]: Ambiguous error message for running `yarn` in nested project
### Self-service

- [ ] I'd be willing to implement a fix

### Describe the bug

Ambiguous error message for running `yarn` in nested project

### To reproduce

```sh
~> mkdir /tmp/a
~> cd /tmp/a
/tmp/a> yarn init -2
/tmp/a> mkdir b
/tmp/a> cd b
/tmp/a/b> npm init -y
/tmp/a/b> yarn
Usage Error: The project in /tmp/a/package.json doesn't seem to have been installed - running an install there might help

$ yarn install [--json] [--immutable] [--immutable-cache] [--check-cache] [--inline-builds] [--mode #0]
```

### Environment

```shell
System:
    OS: Linux 5.13 Manjaro Linux
    CPU: N/A
  Binaries:
    Node: 16.11.1 - /tmp/xfs-55aacde0/node
    Yarn: 3.1.0 - /tmp/xfs-55aacde0/yarn
    npm: 8.1.3 - /usr/bin/npm
```

### Additional context

The following error message is useful:
```
/tmp/a/b> yarn init -2
Usage Error: Cannot use the --install flag from within a project subdirectory

$ yarn init [-p,--private] [-w,--workspace] [-i,--install]
```

Ideally this error message should be used instead of the current error message (running an install there might help).

### PR DESCRIPTION
**Pull request by cuyl**
**What's the problem this PR addresses?**

findPackageLocator or findPackageLocation may throw an Error to break the for-loop.

Closes #3777

**How did you fix it?**

catch the error to continue

**Checklist**
- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [ ] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

**Comment by arcanis** [comment]
I need more information. What error do you have? As far as I can tell, exceptions that occur during this stage are proper errors and shouldn't lead to a successful install.

**Comment by cuyl** [comment]
yarn install in a subfolder contains package.json but not in workspaces.
It should throw 
```
The nearest package directory (/project/temp/package.json) doesn't seem to be part of the project declared in /project.

- If the project directory is right, it might be that you forgot to list /project/temp as a workspace.
- If it isn't, it's likely because you have a yarn.lock or package.json file there, confusing the project root detection.
```
But got 
```
The project in /project/package.json doesn't seem to have been installed - running an install there might help
```

File: packages/yarnpkg-core/sources/Project.ts [hunk 1]
```diff
@@ -624,18 +624,20 @@ export class Project {
...
+            const location = await linker.findPackageLocation(locator, linkerOptions);
+            if (location.replace(TRAILING_SLASH_REGEXP, ``) !== cwd.replace(TRAILING_SLASH_REGEXP, ``)) {
+              continue;
+            }
           }
+          return locator;
         }
-        return locator;
-      }
+      } catch (_) {}
```
Comment (paul-soporan) [inline packages/yarnpkg-core/sources/Project.ts:640]:
```suggestion
      } catch {}
```

**Comment by merceyz** [comment]
Please add a test to https://github.com/yarnpkg/berry/blob/d5454007c9c76becfa97b36a92de299a3694afd5/packages/acceptance-tests/pkg-tests-specs/sources/commands/install.test.js that makes sure your changes fixes the issue

**Comment by cuyl** [comment]
@merceyz test added

**Comment by paul-soporan** [comment]
> Please add a test to [`d545400`/packages/acceptance-tests/pkg-tests-specs/sources/commands/install.test.js](https://github.com/yarnpkg/berry/blob/d5454007c9c76becfa97b36a92de299a3694afd5/packages/acceptance-tests/pkg-tests-specs/sources/commands/install.test.js) that makes sure your changes fixes the issue

@merceyz Wouldn't it make more sense for the test to be added to https://github.com/yarnpkg/berry/blob/master/packages/acceptance-tests/pkg-tests-specs/sources/features/projectDetection.test.js since the changes affect `Project.find` which is about project detection and not installs specifically :thinking:

**Comment by arcanis** [comment]
I think this change doesn't work; exceptions thrown by `findPackageLocation` are actual errors, and cannot be skipped safely (otherwise for example, if you remove the `.pnp.cjs`, you'll likely get the "The nearest directory [...]" message, essentially reversing the issue).

I opened #3828 which I think should fix the issue without silencing the errors.

**Pull request closed without merging**